Priority Queue

andrewbriand opened the pull request:
<!--

Thank you for contributing to ___PROJECT___ :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

Adds a GPU-accelerated priority queue

Allows for multiple concurrent insertions as well as multiple concurrent
deletions.

The implementation of the priority queue is based on https://arxiv.org/pdf/1906.06504.pdf.

The queue supports two operations:
   `push`: Add elements into the queue
   `pop`: Remove the element(s) with the lowest (when Max == false) or highest
        (when Max == true) keys

The priority queue supports bulk host-side operations and more fine-grained
device-side operations.

The host-side bulk operations `push` and `pop` allow an arbitrary number of
elements to be pushed to or popped from the queue.

The device-side operations allow a cooperative group to push or pop
some number of elements less than or equal to node_size. These device side
operations are invoked with a trivially-copyable device view,
`device_mutable_view` which can be obtained with the host function 
`get_mutable_device_view` and passed to the device.

Current limitations:
 - Only supports trivially comparable key types
 - Does not support insertion and deletion at the same time
 - Capacity is fixed and the queue does not automatically resize
 - Deletion from the queue is much slower than insertion into the queue due to congestion at the underlying heap's root node

TODO: Port tests to Catch2 and benchmarks to google benchmark

andrewbriand committed:
Initial priority queue commit

andrewbriand committed:
Add priority queue benchmark

andrewbriand committed:
Class comment

andrewbriand committed:
Improve comments and switch to cuco style

GPUtester commented:
Can one of the admins verify this patch?

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,233 @@
+*             keys in the queue, otherwise, pop operations yeild the elements
+*             with the largest keys
+*/
+template <typename Key, typename Value, bool Max = false>
```
I'd think the behavior controlled by `Max` would instead be controlled by the user-provided comparison operator (that can default to `thrust::less`). That would be more analogous to [`std::priority_queue`](https://en.cppreference.com/w/cpp/container/priority_queue).

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,233 @@
+   *               run
+   */
+  void push(Pair<Key, Value> *elements, size_t num_elements,
+            int block_size = 256, int grid_size = 64000,
```
The details of the kernel launch shouldn't need to be exposed in the public interface.

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,233 @@
+   * @param stream The stream in which the underlying GPU operations will be
+   *               run
+   */
+  void push(Pair<Key, Value> *elements, size_t num_elements,
```
Can this be a begin/end iterator instead?

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,233 @@
+   * @param stream The stream in which the underlying GPU operations will be
+   *               run
+   */
+  void pop(Pair<Key, Value> *out, size_t num_elements,
+           int block_size = 512, int grid_size = 32000,
+           bool warp_level = false,
```
Same comments as above. Iterators and grid launch details are internal.

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,233 @@
+namespace cuco {
+
+/*
+* @brief A GPU-accelerated priority queue of key-value pairs
```
Is there a reason for this to be hardcoded for key-value pairs? Can't it be for any trivially copyable type `T`? e.g., with `std::priority_queue` I could have a `std::priority_queue<int>` or a `std::priority_queue<std::pair<int,int>>`.

jrhemstad replied:
Update docs now that this has been updated.

jrhemstad commented:
I reviewed the top level header at this point and gave some thoughts/questions on how to make this a little more generic.

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,233 @@
+*             with the largest keys
+*/
+template <typename Key, typename Value, bool Max = false>
+class priority_queue {
```
Also, this should have a template for an Allocator to allow customizing how the memory is allocated internally.

andrewbriand committed:
Iterators

andrewbriand committed:
Test for iterators with thrust device_vector

Commit:
Add allocator template parameter

andrewbriand committed:
Allocator

andrewbriand committed:
Accept arbitrary comparison

andrewbriand committed:
Accept arbitrary types instead of just pairs

andrewbriand committed:
Remove pq_pair.h

andrewbriand committed:
Start porting priority queue benchmark to gbenchmark

andrewbriand committed:
Finish porting priority queue benchmark to gbenchmark

andrewbriand committed:
Add multiple node sizes to benchmark

andrewbriand committed:
Start porting tests to Catch2

andrewbriand committed:
Prevent block size from being larger than node size

andrewbriand committed:
Continue porting tests to Catch2

andrewbriand committed:
Make generate_element for KVPair generic

andrewbriand committed:
Finish Catch2 tests

andrewbriand committed:
Hide kernel launch details

andrewbriand committed:
Clean up partial deletion code

andrewbriand committed:
Correct test comparisons

andrewbriand committed:
Commenting and cleanup

andrewbriand committed:
Commenting for Compare

andrewbriand committed:
Cleanup, arbitrary number of elements for device API functions

andrewbriand committed:
Formatting

andrewbriand commented:
Hi Jake,

Thank you for your comments! I've done the following to address them:

1. I've changed the queue to accept arbitrary types instead of just pairs, including a custom comparison operator.
2. I've added the option to specify a custom allocator.
3. I've changed the push and pop functions to accept iterators instead of pointers.
4. I've hidden kernel launch details from the user interface.
     - I hard coded block sizes in `priority_queue.inl` similarly to what is done in `static_map`.
     - I hid the internal node size under a boolean template parameter FavorInsertionPerformance allowing the user to sacrifice deletion performance for insertion performance (by decreasing the node size).
     - I've removed warp level insertion and deletion as it didn't offer any performance improvement in my benchmarking. Furthermore, the user could achieve this if necessary by writing a kernel and using the Device-side API.

Additionally, I have ported my tests to Catch2 and my benchmarks to gbenchmark.

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,238 @@
+* @tparam Allocator Allocator defining how memory is allocated internally
+*/
+template <typename T, typename Compare = thrust::less<T>,
+          bool FavorInsertionPerformance = false,
```
Hm, for my edification, can you expand on what this actually does?

andrewbriand replied:
This decreases the size of the nodes in the queue's underlying heap. This favors insertion performance since we can have more independent blocks working to insert the same amount of data. However, it decreases deletion performance since each block removes less data from the queue at once, leading to more contention at the root node. 

Honestly, we could probably do without this option and just always use the larger node size. In my last test on a Tesla K80, setting this option to true only sped up the insertion of 256M elements from about 4.3s to about 3.8s, while slowing down the deletion of 256M elements from 27s to 183s. My original thinking was that there might be some applications where the user inserts a very large amount of data to the queue but is only interested in retrieving the top few elements.

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,238 @@
+  * @param block_size Size of the blocks to calculate storage for
+  * @return The amount of temporary storage required in bytes
+  */
+  int get_shmem_size(int block_size) {
```
When is this needed by a user? i.e., does it need to be a public function?

andrewbriand replied:
This is needed by a user who wishes to use the device view, as the amount of shared memory required by operations depends on the size of the blocks used for the operation. This could be avoided by always requiring the amount of shared memory required for the maximum block size, if that is better.

andrewbriand replied:
Another option would be to have a shared memory object like in CUB. In that case, we could either require that the block size be known at compile time or assume the maximum block size as above.

jrhemstad commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,238 @@
+     * @param temp_storage Pointer to a contiguous section of memory
+     *        large enough to hold get_shmem_size(g.size()) bytes
+     */
+    template <typename CG, typename OutputIt>
+    __device__ void pop(CG const& g, OutputIt first,
+                        OutputIt last, void *temp_storage);
```
Is it possible to make the temp storage be passed at construction of the view instead of at the operation? That would be more canonical with a CUB like pattern, e.g., BlockReduce:
```
    // Specialize BlockReduce for a 1D block of 128 threads on type int
    using BlockReduce =  cub::BlockReduce<int, 128>;
    // Allocate shared memory for BlockReduce
    __shared__ typename BlockReduce::TempStorage temp_storage;
    // Obtain a segment of consecutive items that are blocked across threads
    int thread_data[4];
    ...
    // Compute the block-wide sum for thread0
    int aggregate = BlockReduce(temp_storage).Sum(thread_data);
```

andrewbriand replied:
Currently, the device view is constructed on the host side, as is done in `cuco::static_map`. One possibility would be to have a function that allows the user to set the temporary memory that the device view uses. 

Maybe another option would be to have an initial, immutable device view that is constructed on the host side, and then passed to a kernel. Once in device code, the user could construct a mutable device view from the immutable device view by passing the temp storage.

Does one of those approaches sound good?

jrhemstad commented:
@andrewbriand left a few more minor comments on the top-level header. Great work! This is very close to done.

andrewbriand committed:
Add missing syncs

andrewbriand commented:
@jrhemstad Thank you very much for your comments! I've responded to them above, let me know what you think.

PointKernel commented:
@andrewbriand Can you resolve the conflict in CMake?

andrewbriand committed:
Merge NVIDIA:dev into andrewbriand:dev

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

andrewbriand commented:
> @andrewbriand Can you resolve the conflict in CMake?

@PointKernel Done!

jrhemstad commented:
Hey @andrewbriand thanks again for the PR! Apologies for letting this languish for so long!

@PointKernel has kindly volunteered to take on shepherding this to getting it ready for merging.

PointKernel commented on benchmarks/priority_queue/priority_queue_bench.cu:
```diff
@@ -0,0 +1,87 @@
+#include <cuco/detail/pair.cuh>
+#include <cuco/priority_queue.cuh>
+
+#include <thrust/device_vector.h>
```
```suggestion
#include <cuco/detail/pair.cuh>
#include <cuco/priority_queue.cuh>

#include <thrust/device_vector.h>

#include <benchmark/benchmark.h>

#include <cstdint>
#include <random>
#include <vector>
```
Header groups should be reordered. A general rule is explained [here](https://github.com/NVIDIA/cuCollections/pull/141#discussion_r876111582)

andrewbriand replied:
0196bde

PointKernel commented on benchmarks/priority_queue/priority_queue_bench.cu:
```diff
@@ -0,0 +1,87 @@
+#include <cstdint>
```
```suggestion
/*
 * Copyright (c) 2021-2022, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cstdint>
```
Copyright to be added

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,239 @@
+#include <utility>
+#include <vector>
+
+#include <thrust/functional.h>
```
```suggestion
#include <cuco/allocator.hpp>

#include <thrust/functional.h>

#include <utility>
#include <vector>
```

andrewbriand replied:
0196bde

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,239 @@
+                             ///  lock for the node starting at
+                             ///  1d_heap_[node_size * i]`
+
+  Allocator allocator_;
```
```suggestion
```
Looks like it's never used.

andrewbriand replied:
bf930dd

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,239 @@
+    int* d_size_;
+    size_t* d_p_buffer_size_;
+    int* d_locks_;
+    Compare compare_;
```
Docs

andrewbriand replied:
54dc9f3

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,239 @@
+   * @param initial_capacity The number of elements the priority queue can hold
+   * @param alloc Allocator used for allocating device storage
+   */
+  priority_queue(size_t initial_capacity, Allocator const& alloc = Allocator{});
```
Maybe we should pass a stream parameter to set `d_*` variables asynchronously.

andrewbriand replied:
4269e9c

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,239 @@
+                             ///  1d_heap_[node_size * i]`
+
+  Allocator allocator_;
+  int_allocator_type int_allocator_;
+  t_allocator_type t_allocator_;
+  size_t_allocator_type size_t_allocator_;
```
Docs

andrewbriand replied:
54dc9f3

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,239 @@
+  t_allocator_type t_allocator_;
+  size_t_allocator_type size_t_allocator_;
+
+  Compare compare_{};
```
```suggestion
```
Just for my own education, is it legitimate to use different comparators for push and pop or for multiple pushes?

If so, `Compare` should be a template parameter for public APIs like `push`/`pop` instead of an attribute in `priority_queue` or its view. e.g.
```cpp
template <typename InputIt, typename Compare = thrust::less<T>>
void push(InputIt first, InputIt last, Compare compare = Compare{}, cudaStream_t stream = 0); 
```

andrewbriand replied:
I don't think this would be legitimate, as the heap structure where the elements are stored keeps them ordered according to a certain comparator. In what case might a user want to supply different comparators for different operations?

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,239 @@
+    int* d_size_;
+    size_t* d_p_buffer_size_;
+    int* d_locks_;
+    Compare compare_;
```
```suggestion
```

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,178 @@
+                                                                            InputIt last,
+                                                                            cudaStream_t stream)
+{
+  const int kBlockSize = min(256, (int)node_size_);
```
```suggestion
  const int kBlockSize = std::min(256, static<int>(node_size_));
```
Please avoid c-style expressions.

Actually, there is an issue here: what if node size is not a multiple of warp size and is smaller than 256? For the performance purpose, the block size should always be a multiple of 32. The common choices are 128 and 256. We may want to determine grid/block dim like other cuco implementations (see [here](https://github.com/NVIDIA/cuCollections/blob/edb0bfda0322b3595e56db9db05427401bed732e/include/cuco/detail/static_map.inl#L134-L137))?

andrewbriand replied:
Currently, node size is determined [here](https://github.com/andrewbriand/cuCollections/blob/dev/include/cuco/priority_queue.cuh#L77) according to the template parameter `FavorInsertionPerformance`, and so is always a multiple of 32. Block size depends on node size because any extra threads past the node size in each block will simply sit idle in the internal heap's operations.

I see that in other cuco implementations, block size is usually fixed. However, as discussed briefly with Jake [above](https://github.com/NVIDIA/cuCollections/pull/105#discussion_r848905152), I included `FavorInsertionPerformance` because I was thinking that there might be some applications where the user inserts a very large amount of data to the queue but is only interested in retrieving the top few elements. I now think though that removing this may be better, given that the performance increase to insertion is not that significant (4.3s to 3.8s on a Tesla K80 for 256M elements in my testing), and that it would simplify the code.

PointKernel replied:
The 10% speedup is not trivial but I'm fine with removing `FavorInsertionPerformance` for now. We can get this PR merged first and look into optimization later. Another reason to remove `FavorInsertionPerformance` is that a boolean parameter dedicated to performance tuning is not part of the priority queue characteristics thus should not be exposed to the public IMO. We need to rethink this.

andrewbriand replied:
Ok, I've removed `FavorInsertionPerformance` in 55cf2e6

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,178 @@
+  const int kBlockSize = min(256, (int)node_size_);
+  const int kNumBlocks = min(64000, max(1, (int)((last - first) / node_size_)));
+
+  PushKernel<<<kNumBlocks, kBlockSize, get_shmem_size(kBlockSize), stream>>>(first,
```
`cuco` is basically following [libcudf naming convention](https://github.com/rapidsai/cudf/blob/branch-22.08/cpp/docs/DEVELOPER_GUIDE.md#code-and-documentation-style-and-formatting).

andrewbriand replied:
30cbf83, 192e263, and 9da822f

PointKernel commented on include/cuco/detail/priority_queue_kernels.cuh:
```diff
@@ -0,0 +1,1214 @@
+#include <assert.h>
+#include <cooperative_groups.h>
+
+using namespace cooperative_groups;
+
+namespace cuco {
```
```suggestion
namespace cuco {
namespace detail {
namespace cg = cooperative_groups;
```
All kernels should be placed in `detail` namespace.

andrewbriand replied:
bec63f3

PointKernel commented on benchmarks/priority_queue/priority_queue_bench.cu:
```diff
@@ -0,0 +1,87 @@
+  __host__ __device__ bool operator()(const T& a, const T& b) const { return a.first < b.first; }
+};
+
+template <typename Key, typename Value, typename OutputIt>
+static void generate_keys_uniform(OutputIt output_begin, OutputIt output_end)
+{
+  std::random_device rd;
+  std::mt19937 gen{rd()};
+
+  auto num_keys = std::distance(output_begin, output_end);
+
+  for (auto i = 0; i < num_keys; ++i) {
+    output_begin[i] = {static_cast<Key>(gen()), static_cast<Value>(gen())};
+  }
+}
```
Can this be replaced by `generate_keys` in `benchmarks/key_generator.hpp`?

andrewbriand replied:
This won't work directly, as this function is generating key-value pairs rather than just keys. We could use `generate_keys` to generate both keys and values and then combine them or use it to generate keys and then assign sequential values though. For now, I have renamed `generate_keys_uniform` -> `generate_kv_pairs_uniform` for clarity (aa12404).

PointKernel reviewed (changes requested):
The first round of reviews with mainly coding style comments. I didn't comment on all the files but the suggestions are applied to all files in this PR. A checklist of issues to address in all files:

- [x] Copyright information
- [x] Reorder header groups
- [x] Follow [snake_case](https://en.wikipedia.org/wiki/Snake_case) naming convention for all names except in a few cases: template parameters, unit tests and test case names may use[ Camel case](https://en.wikipedia.org/wiki/Camel_case)
- [x] Avoid c-style code e.g. `std::min` instead of `min`, `std::size_t` instead of `size_t`, `static_cast<int>(size)` instead of `int(size)`, etc. and include corresponding headers
- [x] Use `constexpr` or `const` whenever possible

andrewbriand committed:
Add copyright to priority_queue_bench.cu

Co-authored-by: Yunsong Wang <wangyunsong89@gmail.com>

andrewbriand committed:
Add copyright to priority queue files

andrewbriand committed:
Order headers from near to far in priority queue files

andrewbriand committed:
Bug fix in priority queue test code

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

andrewbriand committed:
Remove unnecessary allocator

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

andrewbriand committed:
Add missing member docs in priority_queue.cuh

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

andrewbriand committed:
Add stream parameter to priority queue ctor

andrewbriand committed:
Snake case in priority queue files

andrewbriand committed:
Put priority queue kernels in detail namespace

andrewbriand committed:
generate_keys_uniform -> generate_kv_pairs_uniform

andrewbriand committed:
Remove FavorInsertionPerformance template parameter

andrewbriand committed:
Default node size 64 -> 1024

andrewbriand committed:
Avoid c-style expressions in priority queue files

andrewbriand committed:
Remove FavorInsertionPerformance in priority queue benchmark

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

PointKernel commented:
ok to test

andrewbriand committed:
Snake case in priority_queue_test.cu

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

andrewbriand committed:
kPBufferIdx -> p_buffer_idx and kRootIdx -> root_idx

andrewbriand committed:
Use const and constexpr wherever possible in priority queue files

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

andrewbriand commented:
@PointKernel Thanks for your comments! I believe that I have addressed or responded to them all. Please let me know what you think and what other comments you might have.

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,264 @@
+   * @brief Construct a priority queue
+   *
+   * @param initial_capacity The number of elements the priority queue can hold
+   * @param alloc Allocator used for allocating device storage
```
```suggestion
   * @param alloc Allocator used for allocating device storage
   * @param stream Stream used for constructing the priority queue
```

andrewbriand replied:
7c4b1f6

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,264 @@
+   * @param block_size Size of the blocks to calculate storage for
+   * @return The amount of temporary storage required in bytes
+   */
+  int get_shmem_size(int block_size) const
```
```suggestion
  int get_shmem_size(int const block_size) const
```

andrewbriand replied:
1932418

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,264 @@
+   *
+   * @return A device view
+   */
+  device_mutable_view get_mutable_device_view()
```
```suggestion
  device_mutable_view get_mutable_device_view() const noexcept
```

andrewbriand replied:
1932418

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,264 @@
+  ~priority_queue();
+
+  class device_mutable_view {
+   public:
```
```suggestion
   public:
     using value_type = T;
```

andrewbriand replied:
Should I also replace references to `T` with `value_type` in `device_mutable_view`?

PointKernel replied:
That will be great!

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,197 @@
+  const int num_nodes  = static_cast<int>((last - first) / node_size_) + 1;
+  const int num_blocks = std::min(64000, num_nodes);
+
+  detail::push_kernel<<<num_blocks, block_size, get_shmem_size(block_size), stream>>>(
+    first,
+    last - first,
+    d_heap_,
+    d_size_,
+    node_size_,
+    d_locks_,
+    d_p_buffer_size_,
+    lowest_level_start_,
+    compare_);
```
```suggestion
  auto view = get_device_mutable_view();
  detail::push_kernel<<<num_blocks, block_size, get_shmem_size(block_size), stream>>>(
    first, num_elements, view);
```
This is a great example showing the power of "view". Accordingly, the `push_kernel` would look like:
```cpp
template <typename OutputIt, typename viewT>
__global__ void push_kernel(OutputIt elements,
                            std::size_t const num_elements,
                            viewT view)
{
  using T = typename viewT::value_type;
  ...
}
``` 
If you want, `push_n_kernel` instead of `push_kernel` would be a more descriptive name in this case.

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,197 @@
+  detail::pop_kernel<<<num_blocks, block_size, get_shmem_size(block_size), stream>>>(
+    first,
+    pop_size,
+    d_heap_,
+    d_size_,
+    node_size_,
+    d_locks_,
+    d_p_buffer_size_,
+    lowest_level_start_,
+    node_capacity_,
+    compare_);
```
Use `view` instead

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,197 @@
+{
+  constexpr int block_size = 256;
+
+  const int num_nodes  = static_cast<int>((last - first) / node_size_) + 1;
```
```suggestion
  auto const num_elements = std::distance(first, last);
  const int num_nodes  = static_cast<int>(num_elements / node_size_) + 1;
```
`last - first` is not always safe for iterators.

PointKernel replied:
Please note the empty input should be properly handled wrt #151

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,197 @@
+  constexpr int block_size = 256;
+
+  const int num_nodes  = static_cast<int>((last - first) / node_size_) + 1;
+  const int num_blocks = std::min(64000, num_nodes);
```
Can we avoid using the magic number `64000`?

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,197 @@
+void priority_queue<T, Compare, Allocator>::pop(OutputIt first, OutputIt last, cudaStream_t stream)
+{
+  constexpr int block_size = 256;
+  const int pop_size       = last - first;
```
`std::distance`

PointKernel commented on include/cuco/detail/priority_queue.inl:
```diff
@@ -0,0 +1,197 @@
+
+  const auto push_size = last - first;
+  for (std::size_t i = 0; i < push_size / node_size_; i++) {
+    detail::push_single_node(g,
+                             first + i * node_size_,
+                             d_heap_,
+                             d_size_,
+                             node_size_,
+                             d_locks_,
+                             lowest_level_start_,
+                             shmem,
+                             compare_);
```
`push_single_node`, `push_partial_node`, and related utilities should be member functions of `device_mutable_view`.

PointKernel replied:
The same as `pop_single_node` and `pop_partial_node`

PointKernel commented on include/cuco/priority_queue.cuh:
```diff
@@ -0,0 +1,264 @@
+    template <typename CG, typename OutputIt>
+    __device__ void pop(CG const& g, OutputIt first, OutputIt last, void* temp_storage);
+
+    /*
+     * @brief Return the amount of temporary storage required for operations
+     * on the queue with a cooperative group size of block_size
+     *
+     * @param block_size Size of the cooperative groups to calculate storage for
+     * @return The amount of temporary storage required in bytes
+     */
+    __device__ int get_shmem_size(int block_size) const
+    {
+      int intersection_bytes = 2 * (block_size + 1) * sizeof(int);
+      int node_bytes         = node_size_ * sizeof(T);
+      return intersection_bytes + 2 * node_bytes;
+    }
```
This seems never used

PointKernel commented on include/cuco/detail/priority_queue_kernels.cuh:
```diff
@@ -0,0 +1,1236 @@
+ * @param shmem The shared memory layout for this cooperative group
+ * @param compare Comparison operator ordering the elements in the heap
+ */
+template <typename InputIt, typename T, typename Compare, typename CG>
```
`OutputIt` instead of `InputIt`

PointKernel commented on include/cuco/detail/priority_queue_kernels.cuh:
```diff
@@ -0,0 +1,1236 @@
+* @param lowest_level_start The first index of the heaps lowest layer
+* @param compare Comparison operator ordering the elements in the heap
+*/
+template <typename OutputIt, typename T, typename Compare>
```
```suggestion
template <typename InputIt, typename viewT>
```

PointKernel commented on include/cuco/detail/priority_queue_kernels.cuh:
```diff
@@ -0,0 +1,1236 @@
+template <typename OutputIt, typename T, typename Compare>
+__global__ void push_kernel(OutputIt elements,
+                            std::size_t num_elements,
+                            T* heap,
+                            int* size,
+                            std::size_t node_size,
+                            int* locks,
+                            std::size_t* p_buffer_size,
+                            int lowest_level_start,
+                            Compare compare)
```
```suggestion
                            viewT view)
```

The kernel implementation can also be simplified with view.

PointKernel reviewed (changes requested):
Another round of review.

Thanks @andrewbriand for your effort and persistence made to this PR! We are almost there.

andrewbriand committed:
Add missing const in priority queue

Co-authored-by: Yunsong Wang <wangyunsong89@gmail.com>

andrewbriand committed:
Add docs for stream parameter to priority queue ctor

Co-authored-by: Yunsong Wang <wangyunsong89@gmail.com>

andrewbriand committed:
Add value_type to priority_queue::device_mutable_view

Co-authored-by: Yunsong Wang <wangyunsong89@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

PointKernel commented:
@andrewbriand Can you please also merge with the latest `dev` branch and fix build warnings (if there is any)?

arsdever commented:
Hi guys. Is this gonna make it into the dev branch?

PointKernel commented:
> Hi guys. Is this gonna make it into the dev branch?

@arsdever Thanks for following up. This work is currently on our backlog, and we’ll let you know once we have a more definite timeline for completion.
